src/Disks/ObjectStorages/MetadataStorageFromPlainObjectStorageOperations.h

```diff
@@ -0,0 +1,80 @@
+#pragma once
+
+#include <Disks/ObjectStorages/IMetadataOperation.h>
+#include <Disks/ObjectStorages/MetadataStorageFromPlainObjectStorage.h>
+
+#include <filesystem>
+#include <map>
+
+namespace DB
+{
+
+class MetadataStorageFromPlainObjectStorageCreateDirectoryOperation final : public IMetadataOperation
+{
+private:
+    std::filesystem::path path;
+    std::string key_prefix;
+    MetadataStorageFromPlainObjectStorage::PathMap & path_map;
+    ObjectStoragePtr object_storage;
+
+    bool write_created = false;
+    bool write_finalized = false;
+
+public:
+    // Assuming that paths are normalized.
+    MetadataStorageFromPlainObjectStorageCreateDirectoryOperation(
+        std::filesystem::path && path_,
+        std::string && key_prefix_,
+        MetadataStorageFromPlainObjectStorage::PathMap & path_map_,
+        ObjectStoragePtr object_storage_);
+
+    void execute(std::unique_lock<SharedMutex> & metadata_lock) override;
+    void undo() override;
```

Comment:
I agree the lock reference should be part of the `undo` signature.
I was reluctant to change it since that will also require the change of multiple `MetadataStorageFromDiskTransactionOperations` after I extracted the common interface: https://github.com/ClickHouse/ClickHouse/blob/master/src/Disks/ObjectStorages/MetadataStorageFromDiskTransactionOperations.h

There is no bug because `undo` is done [under the lock](https://github.com/ClickHouse/ClickHouse/pull/61116/files#diff-2b99730b3a5ca3f8a55c53fc2789914c93c90da4a7fed49fc2a604d8dd3e3232R60)
Since the current implementation is confusing I'll change it, should be a purely cosmetic change.